### PR TITLE
Audit log export

### LINKED ISSUE #3992: Allow csv export of audit-log
### Epic

_No response_

### Description

**As a:** team owner

**I want to:** be able to download a copy of the audit log locally for auditing/recording

**So that:** I can meet my audit requirements

---

Proposal is a button to download the audit log in csv format.

https://app-eu1.hubspot.com/contacts/26586079/record/0-2/9822854636

### Which customers would this be available to

Everyone - CE/Starter/Team/Enterprise

### Acceptance Criteria

_No response_

### Have you provided an initial effort estimate for this issue?

I have provided an initial effort estimate

### PR DESCRIPTION
**Pull request by hardillb**
fixes #3992
## Description

Adds the following api endpoints
- /api/v1/applications/{id}/audit-log/export
- /api/v1/projects/{id}/audit-log/export
- /api/v1/teams/{id}/audit-log/export
- /api/v1/admin/{id}/audit-log/export

All output CSV similar to

```
id,event,body,scope,trigger,createdAt
pqGOpkz15V,project.deleted,"{\"team\":{\"id\":\"32E4NEO5pY\",\"name\":\"Development\",\"slug\":\"development\",\"type\":null},\"project\":{\"id\":\"33d85d33-9ad0-403d-bd65-67e622122567\",\"name\":\"last\"}}","{\"id\":\"32E4NEO5pY\",\"type\":\"team\"}","{\"id\":\"DMpxaBOzEv\",\"type\":\"user\",\"name\":\"alice\"}","Fri Aug 02 2024 11:19:51 GMT+0100 (British Summer Time)"
2LGD8QrGwn,project.created,"{\"team\":{\"id\":\"32E4NEO5pY\",\"name\":\"Development\",\"slug\":\"development\",\"type\":null},\"project\":{\"id\":\"33d85d33-9ad0-403d-bd65-67e622122567\",\"name\":\"last\"}}","{\"id\":\"32E4NEO5pY\",\"type\":\"team\"}","{\"id\":\"DMpxaBOzEv\",\"type\":\"user\",\"name\":\"alice\"}","Fri Aug 02 2024 11:18:43 GMT+0100 (British Summer Time)"
Lq1Nlj7x6j,application.pipeline.stage-deployed,"{\"team\":{\"id\":\"32E4NEO5pY\",\"name\":\"Development\",\"slug\":\"development\",\"type\":null},\"application\":{\"id\":5,\"name\":\"foo\"},\"pipeline\":{\"id\":\"nALXBw4GO5\",\"name\":\"locked\"},\"pipelineStage\":{\"id\":\"152fb007-7eb4-4e41-ad3b-8f946d11108b\",\"name\":\"first\"},\"pipelineStageTarget\":{\"id\":\"d04494fa-0fcc-42e1-a56f-76acab1d87cb\",\"name\":\"fourth\"}}","{\"id\":\"32E4NEO5pY\",\"type\":\"team\"}","{\"id\":\"DMpxaBOzEv\",\"type\":\"user\",\"name\":\"alice\"}","Wed Jul 24 2024 14:45:23 GMT+0100 (British Summer Time)"
6Qm4Je6GJX,application.pipeline.stage-deployed,"{\"team\":{\"id\":\"32E4NEO5pY\",\"name\":\"Development\",\"slug\":\"development\",\"type\":null},\"application\":{\"id\":5,\"name\":\"foo\"},\"pipeline\":{\"id\":\"nALXBw4GO5\",\"name\":\"locked\"},\"pipelineStage\":{\"id\":\"152fb007-7eb4-4e41-ad3b-8f946d11108b\",\"name\":\"first\"},\"pipelineStageTarget\":{\"id\":\"d04494fa-0fcc-42e1-a56f-76acab1d87cb\",\"name\":\"fourth\"}}","{\"id\":\"32E4NEO5pY\",\"type\":\"team\"}","{\"id\":\"DMpxaBOzEv\",\"type\":\"user\",\"name\":\"alice\"}","Wed Jul 24 2024 14:43:30 GMT+0100 (British Summer Time)"
pqGOp3b15V,application.pipeline.stage-deployed,"{\"team\":{\"id\":\"32E4NEO5pY\",\"name\":\"Development\",\"slug\":\"development\",\"type\":null},\"application\":{\"id\":5,\"name\":\"foo\"},\"pipeline\":{\"id\":\"nALXBw4GO5\",\"name\":\"locked\"},\"pipelineStage\":{\"id\":\"152fb007-7eb4-4e41-ad3b-8f946d11108b\",\"name\":\"first\"},\"pipelineStageTarget\":{\"id\":\"d04494fa-0fcc-42e1-a56f-76acab1d87cb\",\"name\":\"fourth\"}}","{\"id\":\"32E4NEO5pY\",\"type\":\"team\"}","{\"id\":\"DMpxaBOzEv\",\"type\":\"user\",\"name\":\"alice\"}","Wed Jul 24 2024 14:32:48 GMT+0100 (British Summer Time)"
```

## Related Issue(s)

#3992

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/FlowFuse/flowfuse/blob/main/CONTRIBUTING.md)
 - [ ] Suitable unit/system level tests have been added and they pass 
 - [x] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `FlowFuse/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `FlowFuse/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

**Review by Steve-Mcl — COMMENTED** [review]
### Invalid CSV

Unfortunately, stringifying an object (even double stringify) will not create valid RFC4180 CSV.  Additionally, RFC states line endings should be `\r\n`

For example, this implementation will output:

```
id,event,body,scope,trigger,createdAt\n
mzgb4KvD20,account.login,,"{\"id\":\"EexY4j17B2\",\"type\":\"user\"}","{\"id\":\"EexY4j17B2\",\"type\":\"user\",\"name\":\"steve\"}","Wed Aug 07 2024 14:21:22 GMT+0100 (British Summer Time)"\n
aZ8KXLkDlz,account.login,,"{\"id\":\"6J5Ybp1pbr\",\"type\":\"user\"}","{\"id\":\"6J5Ybp1pbr\",\"type\":\"user\",\"name\":\"member\"}","Wed Jul 31 2024 08:16:00 GMT+0100 (British Summer Time)"\n

```

For it to be valid CSV it should be:

```
id,event,body,scope,trigger,createdAt\r\n
mzgb4KvD20,account.login,,"{""id"":""EexY4j17B2"",""type"":""user""}","{""id"":""EexY4j17B2"",""type"":""user"",""name"":""steve""}","Wed Aug 07 2024 14:21:22 GMT+0100 (British Summer Time)"\r\n
aZ8KXLkDlz,account.login,,"{""id"":""6J5Ybp1pbr"",""type"":""user""}","{""id"":""6J5Ybp1pbr"",""type"":""user"",""name"":""member""}","Wed Jul 31 2024 08:16:00 GMT+0100 (British Summer Time)"\r\n
```

NOTES:
* RFC4180 2.1 states: lines are separated by CRLF
* RFC4180 2.7 states: double-quote appearing inside a field must be escaped by preceding it with another double quote


<br>


### No Unit tests (definition of done)
I dont see any UNIT tests defined for the new routes?  for example checking raw IDs are not leaked.   As an example, this code in its current implementation would leak user ID for event `users.created-user`:

```
rWjYXnX83d,users.created-user,"{\"user\":{\"id\":77,\"name\":\"viewer\",\"username\":\"viewer\",\"email\":\"viewer@home.local\"}}","{\"id\":\"7GdMDZg3ja\",\"type\":\"user\"}","{\"id\":\"EexY4j17B2\",\"type\":\"user\",\"name\":\"steve\"}","Tue Feb 13 2024 10:18:00 GMT+0000 (Greenwich Mean Time)"
```

Though to be completely fair, this is probably better attributed to being an existing BUG (will raise issue shortly). However regressions of this nature would be missed without tests.


<br>

### For consideration: 

`createdAt` field should probably be exported in ISO 8601 format if the data is intended to be ingested by other systems.  Current implementation will result in local format e.g. `Wed Jul 31 2024 08:16:00 GMT+0100 (British Summer Time)`. A more universally accepted format might be preferable in the 1st iteration.

Also, this field does not need to be quoted.


<br>


### For consideration: 

Lastly, a note regarding "Adding new routes" -vs- "Reusing existing routes" 

Some might argue the existing route should make use of something like `Accept: text/csv` header however I will not block on this. FWIW, I personally prefer the clear separation of concerns separate routes provide. Will leave this as a consideration point only.

**Comment by Steve-Mcl** [comment]
> [x] Documentation has been updated

Did you forget to push docs update Ben?

**Comment by hardillb** [comment]
@Steve-Mcl the doc update was done by including the swager comment before the new functions, so the API doc has been updated.

**Comment by Steve-Mcl** [comment]
> @Steve-Mcl the doc update was done by including the swager comment before the new functions, so the API doc has been updated.

No bother. Just with it being ticked I thought perhaps you had some example curls or info sitting un-pushed on your local dev.

**Comment by hardillb** [comment]
@Steve-Mcl 

The timestamp is what we get back from the auditlog sanitizer (the thing that should be converting all the ids to hashes), we would need to re-parse them to convert them back to timestamps.

Please try the new CSV escaping

File: forge/routes/api/admin.js [hunk 1]
```diff
@@ -303,6 +303,56 @@ module.exports = async function (app) {
...
+        const result = app.db.views.AuditLog.auditLog(logEntries)
+        reply.type('text/csv').send([
+            ['id', 'event', 'body', 'scope', 'trigger', 'createdAt'],
+            ...result.log.map(row => [
+                row.id,
+                row.event,
+                `"${row.body ? JSON.stringify(row.body).replace(/"/g, '""') : ''}"`,
+                `"${JSON.stringify(row.scope).replace(/"/g, '""')}"`,
+                `"${JSON.stringify(row.trigger).replace(/"/g, '""')}"`,
+                row.createdAt
```
Comment (Steve-Mcl) [inline forge/routes/api/admin.js:349]:
```suggestion
                row.createdAt?.toISOString()
```

File: forge/routes/api/application.js [hunk 2]
```diff
@@ -477,4 +477,60 @@ module.exports = async function (app) {
...
+        const result = app.db.views.AuditLog.auditLog(logEntries)
+        reply.type('text/csv').send([
+            ['id', 'event', 'body', 'scope', 'trigger', 'createdAt'],
+            ...result.log.map(row => [
+                row.id,
+                row.event,
+                `"${row.body ? JSON.stringify(row.body).replace(/"/g, '""') : ''}"`,
+                `"${JSON.stringify(row.scope).replace(/"/g, '""')}"`,
+                `"${JSON.stringify(row.trigger).replace(/"/g, '""')}"`,
+                row.createdAt
```
Comment (Steve-Mcl) [inline forge/routes/api/application.js:530]:
```suggestion
                row.createdAt?.toISOString()
```

File: forge/routes/api/device.js [hunk 3]
```diff
@@ -942,6 +942,61 @@ module.exports = async function (app) {
...
+        const result = app.db.views.AuditLog.auditLog(logEntries)
+        reply.type('text/csv').send([
+            ['id', 'event', 'body', 'scope', 'trigger', 'createdAt'],
+            ...result.log.map(row => [
+                row.id,
+                row.event,
+                `"${row.body ? JSON.stringify(row.body).replace(/"/g, '""') : ''}"`,
+                `"${JSON.stringify(row.scope).replace(/"/g, '""')}"`,
+                `"${JSON.stringify(row.trigger).replace(/"/g, '""')}"`,
+                row.createdAt
```
Comment (Steve-Mcl) [inline forge/routes/api/device.js:993]:
```suggestion
                row.createdAt?.toISOString()
```

File: forge/routes/api/project.js [hunk 4]
```diff
@@ -1034,6 +1034,62 @@ module.exports = async function (app) {
...
+        const result = app.db.views.AuditLog.auditLog(logEntries)
+        reply.type('text/csv').send([
+            ['id', 'event', 'body', 'scope', 'trigger', 'createdAt'],
+            ...result.log.map(row => [
+                row.id,
+                row.event,
+                `"${row.body ? JSON.stringify(row.body).replace(/"/g, '""') : ''}"`,
+                `"${JSON.stringify(row.scope).replace(/"/g, '""')}"`,
+                `"${JSON.stringify(row.trigger).replace(/"/g, '""')}"`,
+                row.createdAt
```
Comment (Steve-Mcl) [inline forge/routes/api/project.js:1087]:
```suggestion
                row.createdAt?.toISOString()
```

File: forge/routes/api/team.js [hunk 5]
```diff
@@ -737,4 +737,60 @@ module.exports = async function (app) {
...
+        const result = app.db.views.AuditLog.auditLog(logEntries)
+        reply.type('text/csv').send([
+            ['id', 'event', 'body', 'scope', 'trigger', 'createdAt'],
+            ...result.log.map(row => [
+                row.id,
+                row.event,
+                `"${row.body ? JSON.stringify(row.body).replace(/"/g, '""') : ''}"`,
+                `"${JSON.stringify(row.scope).replace(/"/g, '""')}"`,
+                `"${JSON.stringify(row.trigger).replace(/"/g, '""')}"`,
+                row.createdAt
```
Comment (Steve-Mcl) [inline forge/routes/api/team.js:790]:
```suggestion
                row.createdAt?.toISOString()
```

**Review by Steve-Mcl — COMMENTED** [review]
The time stamp is from the database Ben. There is no reparsing (under the hood `.toString()` is called implicitly, by adding `?.toISOString()` there is no additional processing.

Also, doing this prevents cases where the "friendly time string" might contain commas (almost certain i've seen `toString` output `Pacific Time - Los Angeles, San Diego` which would break the CSV with mismatched row/header count)

**Review by Steve-Mcl — APPROVED** [review]
Looking good.

image

**Pull request merged**